### ISSUE TITLE
Discussion of Redis pub/sub mechanism for scaling Yjs websocket service 

### ISSUE THREAD

**Issue by sbavoma** [kapv89/yjs-scalable-ws-backend#11]
Hi @kapv89, first of all great implementation it helps in making yjs stateless.

I have few questions and findings which I think we can discuss here:
 
 1. I implemented the same redis implementation as in master but I see the awareness handler will cause an infinite loop cause it publishes its awareness in queue and the subscriber listens and applies the update (this update will again have awareness loop)
    check lines 
    https://github.com/kapv89/yjs-scalable-ws-backend/blob/main/src/setupWSConnection.ts#L139
    https://github.com/kapv89/yjs-scalable-ws-backend/blob/main/src/setupWSConnection.ts#L296
 I think for update handler also a loop can form which will raise concern

 2. The check to publish the update in queue is to check the connection type to be of websocket and if its from the same server instance where it is going to be published
    https://github.com/kapv89/yjs-scalable-ws-backend/blob/main/src/setupWSConnection.ts#L228
    
    Lets say if I have my service running behind loadbalancer, then in that case no connection will have origin as a valid check and hence it can fail. Can you share more on this as in what cases the check will work

**Comment by kapv89** [comment]
1. `messageListener` is hooked here - https://github.com/kapv89/yjs-scalable-ws-backend/blob/main/src/setupWSConnection.ts#L50 ... specifically on websocket message ...  `applyAwarenessUpdate` in the redis channel subscriber triggers this - https://github.com/kapv89/yjs-scalable-ws-backend/blob/main/src/setupWSConnection.ts#L50 ... no infinite loop here.
2. Even if your service is behind a load balancer, the load balancer will still open a websocket to your service, and that websocket will be the origin of your updates ... the second check `doc.conns.has(origin)` is technically not really needed, i just found it cleaner

**Comment by sbavoma** [comment]
Hi @kapv89 , thank you for the reply.
I see erdtool is live, seems interesting I'll give it a try :)
So I was testing the implementation of yjs scalable backend and I saw few things
For 10 concurrent clients using 2 servers and same redis I was seeing OOM crashes (~2GB RAM consumed )
Now when I am debugging on local it seems the redis queue is always > 100 in length and there are plenty of read-writes increasing exponentially in order of increased clients 

Need some help on understanding the redis publishBuffer/pushBuffer part and how to test the implementation for no of connections, system metrics profiling etc.

**Comment by kapv89** [comment]
Interesting .. would you be willing to hop on a call sometime?

@sbavoma .. what sort of load were you putting  on the system with those 10 clients?

**Comment by sbavoma** [comment]
yup @kapv89 anytime around evening is fine. I use discord for open source projects ping me sbalikondwar#8756
For these 10 clients I opened one document and was just editing those so ideally should have very less memory footprint but it does not.
On futher debugging I found that every update by one client calls updateHandler in all other clients, so for 10 clients which are simultaneously updating doc, they will have 10 updates done at expense of 1 update

**Comment by sbavoma** [comment]
Greetings @kapv89 
I am finally done with my changes for scaling our service by referring your implementation
Services seems to work fine although had to do lot of changes since our application was a bit different.
At the moment application seems to handle good traffic with 200+ connections. 
Only one thing we have observed is after a while the machine fails to free up RAM and is consuming significant amount of it.
First I checked If we were unsubscribing the channels which we do hence I am a bit puzzled where this can happen.
Once again, nice work! 🥳 

Also I have my load test script I will create a commit to this when i get free.
